Milddleware process_resource() requires four arguments

Undocumented by http://falcon.readthedocs.org/en/latest/api/middleware.html a middleware component's `process_resource()` method requires a fourth argument which is the `params`.

I think I ran into this issue as well. I tried to get the [more complex example](https://falcon.readthedocs.org/en/stable/user/quickstart.html#more-features) working and encountered the following error on any request:

```
2016-04-16 18:23:44 +0200] [13566] [ERROR] Error handling request /
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/gunicorn/workers/sync.py", line 130, in handle
    self.handle_request(listener, req, client, addr)
  File "/Library/Python/2.7/site-packages/gunicorn/workers/sync.py", line 171, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/Library/Python/2.7/site-packages/falcon/api.py", line 167, in __call__
    self._call_req_mw(middleware_stack, req, resp)
  File "/Library/Python/2.7/site-packages/falcon/api.py", line 535, in _call_req_mw
    process_request(req, resp)
  File "/Users/gigi/Development/newspaper3k-api-using-falcon/more-complex-example.py", line 61, in process_request
    href='http://docs.example.com/auth')
TypeError: __init__() takes exactly 4 arguments (3 given)
```

Any hints on how I might resolve this?

--- a/doc/api/middleware.rst
+++ b/doc/api/middleware.rst
@@ -27,7 +27,7 @@ Falcon's middleware interface is defined as follows:
                     the on_* responder.
             """
 
-        def process_resource(self, req, resp, resource):
+        def process_resource(self, req, resp, resource, params):
             """Process the request after routing.
 
             Note:
@@ -41,6 +41,10 @@ Falcon's middleware interface is defined as follows:
                     responder.
                 resource: Resource object to which the request was
                     routed.
+                params: A dict-like object representing any additional
+                    params derived from the route's URI template fields,
+                    that will be passed to the resource's responder
+                    method as keyword arguments.
             """
 
         def process_response(self, req, resp, resource):

--- a/falcon/api.py
+++ b/falcon/api.py
@@ -50,7 +50,7 @@ class API(object):
                                 the on_* responder.
                         \"""
 
-                    def process_resource(self, req, resp, resource):
+                    def process_resource(self, req, resp, resource, params):
                         \"""Process the request and resource *after* routing.
 
                         Note:
@@ -65,6 +65,11 @@ class API(object):
                             resource: Resource object to which the request was
                                 routed. May be None if no route was found for
                                 the request.
+                            params: A dict-like object representing any
+                                additional params derived from the route's URI
+                                template fields, that will be passed to the
+                                resource's responder method as keyword
+                                arguments.
                         \"""
 
                     def process_response(self, req, resp, resource)

